### ISSUE TITLE
Query variable definitions are dropped when used in a nested query

### ISSUE THREAD

**Issue by sambs** [apollostack/apollo-client#620]
It seems that if query variables are used in a nested way such as this:

```
query MyQuery($search:String!) {
  spaces(query:{search:$search}) {
    count
  }
}
```

...then they're dropped from the executed query which ends up looking like this:

```
query MyQuery {
  spaces(query:{search:$search}) {
    count
  }
}
```

**Comment by sambs** [comment]
I've tried rolling back version by version and this issue was introduced in 0.4.12.

**Comment by helfer** [comment]
@sambs oh, interesting. Thanks! Can you make a PR with a failing test for this? That would help getting it fixed faster.

**Comment by stubailo** [comment]
This is because of the traversal we built for #518, but #615 would remove the need for it altogether.

**Comment by danielgriese** [comment]
Hi there,

Any idea when this would be resolved?

While evaluating Apollo for our products, I encountered the same problem by trying to use parameters lower in the query tree (which is totally fine by GraphQL)...

```
query getLabels($account: String!, $search: String) {
  account(id: $account) {
    labels(search: $search) {
      total
      cursor
      items {
        id
        name
      }
    }
  }
}
```

This resulted in the issue that Apollo did only send _query getLabels($account: String!)_  and omitted the $search param, which then resulted in an issue from the server saying that $search is not a defined as a variable...

I have a workaround in place including a dummy "root" query that accepts all the variables that I might need. This works, but I don't think I need to tell you who dirty this feels ;) And I most certainly don't want to push this too production now...

```
query getLabels($account: String!, $search: String) {
  account(id: $account) {
    labels(search: $search) {
    ...
  }
  
  dummy(search: $search)
}
```

Thanks and kind regards,
Daniel

**Comment by helfer** [comment]
@danielgriese We're going to remove query diffing pretty soon (timeframe: less than a week). I'll make sure we take care of this bug as well when we do.

**Comment by stubailo** [comment]
Discussion here: #615, implementation started here: #693

**Comment by stubailo** [comment]
Should be fixed in `0.5.0-1`.
